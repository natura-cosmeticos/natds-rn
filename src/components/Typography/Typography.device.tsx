import { storiesOf } from '@storybook/react-native'
import { All, CustomFonts, Variants } from './Typography.stories'

storiesOf('Typography', module)
  .add('All', All)
  .add('Custom Fonts', CustomFonts)
  .add('Variants', Variants)
