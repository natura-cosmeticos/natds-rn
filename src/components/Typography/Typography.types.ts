import { ReactNode } from 'react';
import { StyleProp, TextProps } from 'react-native';

export type TypographyVariants =
  'heading1'
  | 'heading2'
  | 'heading3'
  | 'heading4'
  | 'heading5'
  | 'heading6'
  | 'subtitle1'
  | 'subtitle2'
  | 'body1'
  | 'body2'
  | 'caption'
  | 'overline'

export interface TypographyProps {
  /**
   * Optional ID for testing
   */
  testID?: string
  /**
   * The component content
   */
  children: ReactNode
  /**
   * Overwrites or extends the styles applied to the component.
   */
  style?: StyleProp<TextProps>
  /**
   * Applies the theme typography styles.
   * @default 'heading1'
   */
  variant?: TypographyVariants
}
