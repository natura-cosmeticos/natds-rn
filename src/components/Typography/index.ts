export { Typography } from './Typography.styles';
