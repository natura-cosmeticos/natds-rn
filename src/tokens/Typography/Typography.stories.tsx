import React from 'react';
import { Text, View } from 'react-native';
import { buildTheme } from '../..';
import {
  getFontFamilyBody, getFontFamilyBranding, getFontFamilyCode, getFontFamilyDisplay, getFontFamilyFallback, getFontFamilyHeadline, getFontFamilyPrimary, getFontFamilySecondary,
} from '../../common/themeSelectors';

export const typography = ({ activeTheme, light }) => {
  const mode = light ? 'light' : 'dark';
  const brandTheme = buildTheme(activeTheme, mode);

  if (!brandTheme) return <Text>No typography found for this theme.</Text>;

  return (
    <View style={{ padding: 8 }}>
      <Text style={{ fontFamily: getFontFamilyPrimary(brandTheme) }}>Primary</Text>
      <Text style={{ fontFamily: getFontFamilySecondary(brandTheme) }}>Secondary</Text>
      <Text style={{ fontFamily: getFontFamilyBranding(brandTheme) }}>Branding</Text>
      <Text style={{ fontFamily: getFontFamilyCode(brandTheme) }}>Code</Text>
      <Text style={{ fontFamily: getFontFamilyDisplay(brandTheme) }}>Display</Text>
      <Text style={{ fontFamily: getFontFamilyHeadline(brandTheme) }}>Headline</Text>
      <Text style={{ fontFamily: getFontFamilyBody(brandTheme) }}>Body</Text>
      <Text style={{ fontFamily: getFontFamilyFallback(brandTheme) }}>Fallback</Text>
    </View>
  );
};

export default {
  parameters: {
    componentSubtitle: 'Custom fonts available in the current theme. Read more in the Documentation section.',
  },
  title: 'Tokens|Typography',
};
