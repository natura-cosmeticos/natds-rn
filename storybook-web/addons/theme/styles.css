.theme {
  font-weight: 400;
  line-height: 1.5;
  max-width: 800px;
  font-family: Roboto, Arial, sans-serif;
  color: #555;
  margin: 30px;
}

.theme__item {
  max-width: 600px;
  border-bottom: rgba(0,0,0,.1) 1px solid;
  display: flex;
  align-items: center;
  padding: 0 0 10px;
  margin: 0 0 10px;
}

.theme__name {
  flex: 2 0 auto;
  font-size: 18px;
  text-transform: capitalize;
  padding: 0;
}

.theme__button {
  background: rgba(0,0,0,.3);
  border: 0;
  border-radius: 4px;
  font-size: 1rem;
  color: #fff;
  text-transform: capitalize;
  cursor: pointer;
  padding: 7px 14px;
  margin: 0 10px 0 0;
}

.theme__button:hover {
  background: rgba(0,0,0,.2);
}

.theme__button:active {
  background: rgba(0,0,0,.25);
}

.theme__button--selected,
.theme__button--selected:hover {
  background: rgba(0,0,0,.6);
  cursor: default;
}
