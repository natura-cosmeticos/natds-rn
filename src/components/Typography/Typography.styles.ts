/* eslint-disable max-len */
import styled from 'styled-components/native';

import { getColorHighlight } from '../../common/themeSelectors';
import { TypographyProps } from './Typography.types';

export const Typography = styled.Text<TypographyProps>(({ theme, variant = 'heading1' }) => ({
  color: getColorHighlight(theme),
  fontSize: theme[variant].fontSize,
  fontWeight: theme[variant].fontWeight,
  letterSpacing: theme[variant].letterSpacing,
}));
