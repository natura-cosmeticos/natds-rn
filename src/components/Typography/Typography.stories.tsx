/* eslint-disable max-lines */
import React from 'react'
import { Platform } from 'react-native'
import { ThemeContext } from 'styled-components/native'
import { Typography } from './Typography.styles'
import { Separator, VerticalStoryContainer } from '../../common/HelperComponents/StoryContainer'

const description = () => `
- - -

  ### NOTE:
  This component is available in the following variants:

  - ✅ Heading1
  - ✅ Heading2
  - ✅ Heading3
  - ✅ Heading4
  - ✅ Heading5
  - ✅ Heading6
  - ✅ Body1
  - ✅ Body2
  - ✅ Caption
  - ✅ Overline

  Props of the nativo component are also available.

- - -
`

export default {
  component: Typography,
  parameters: {
    componentSubtitle: 'Typography component for displaying text.',
    docs: {
      extractComponentDescription: description
    }
  },
  title: 'Components|Typography'
}

type TypographyProps = {
  type: string;
  weight: 'bold' | 'regular';
} | {
  type: string;
  weight?: 'bold' | 'regular';
}

export const CustomFonts = () => {
  const isWebPlatform = Platform.OS === 'web'
  const isAndroidPlatform = Platform.OS === 'android'

  return (
    <ThemeContext.Consumer>
      {(theme) => {
        const isBodyTypography = ({ type = 'body', weight = 'regular' }: TypographyProps) => (type === 'body' ? theme.typography[type][weight] : theme.typography[type])
        const isAndroidBodyTypography = ({ type = 'body', weight = 'regular' }: TypographyProps) => (type === 'body' ? theme.asset.font.file[type][weight] : theme.asset.font.file[type])
        const fontByPlatform = ({ type = 'body', weight = 'regular' }: TypographyProps) => (isWebPlatform ? `"${isBodyTypography({ type, weight }).fontFamily}"` : isBodyTypography({ type, weight }).fontFamily)
        const weightByPlatform = ({ type = 'body', weight = 'regular' }: TypographyProps) => (isWebPlatform ? isBodyTypography({ type, weight }).fontWeight : isBodyTypography({ type, weight }).fontWeight.toString())

        const setTypographyByPlatform = ({ type, weight }: TypographyProps) => {
          if (isAndroidPlatform) {
            return isAndroidBodyTypography({ type, weight })
          }

          return fontByPlatform({ type, weight })
        }

        return (
          <VerticalStoryContainer title="Custom">
            <Typography
              variant="heading3"
              style={{
                fontFamily: setTypographyByPlatform({ type: 'display' }),
                fontWeight: isAndroidPlatform ? 'normal' : weightByPlatform({ type: 'display' })
              }}
            >
              Display

            </Typography>
            <Separator />
            <Typography
              variant="heading3"
              style={{
                fontFamily: setTypographyByPlatform({ type: 'headline' }),
                fontWeight: isAndroidPlatform ? 'normal' : weightByPlatform({ type: 'headline' })
              }}
            >
              Headline

            </Typography>
            <Separator />
            <Typography
              variant="heading3"
              style={{
                fontFamily: setTypographyByPlatform({ type: 'body', weight: 'regular' }),
                fontWeight: isAndroidPlatform ? 'normal' : weightByPlatform({ type: 'body', weight: 'regular' })
              }}
            >
              Body Regular

            </Typography>
            <Separator />
            <Typography
              variant="heading3"
              style={{
                fontFamily: setTypographyByPlatform({ type: 'body', weight: 'bold' }),
                fontWeight: isAndroidPlatform ? 'normal' : weightByPlatform({ type: 'body', weight: 'bold' })
              }}
            >
              Body Bold

            </Typography>
          </VerticalStoryContainer>
        )
      }}
    </ThemeContext.Consumer>
  )
}

export const Variants = () => (
  <VerticalStoryContainer title="Variants">
    <Typography variant="heading1" numberOfLines={1}>Heading1</Typography>
    <Separator />
    <Typography variant="heading2" numberOfLines={1}>Heading2</Typography>
    <Separator />
    <Typography variant="heading3">Heading3</Typography>
    <Separator />
    <Typography variant="heading4">Heading4</Typography>
    <Separator />
    <Typography variant="heading5">Heading5</Typography>
    <Separator />
    <Typography variant="heading6">Heading6</Typography>
    <Separator />
    <Typography variant="body1">Body1</Typography>
    <Separator />
    <Typography variant="body1">Body2</Typography>
    <Separator />
    <Typography variant="caption">Caption</Typography>
    <Separator />
    <Typography variant="overline">Overline</Typography>
  </VerticalStoryContainer>
)
