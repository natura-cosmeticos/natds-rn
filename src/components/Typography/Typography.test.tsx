import React from 'react';
import { Typography } from './Typography.styles';
import { renderWithTheme } from '../../../test/testHelpers';

jest.mock('../../common/themeSelectors', () => (
  {
    getColorHighlight: () => '#000000',
  }));

describe('Typography Component', () => {
  it('should render with default variant', () => {
    const { getByTestId, toJSON } = renderWithTheme(<Typography testID='typography'>Typography</Typography>);

    expect(getByTestId('typography')).toHaveStyle({ fontSize: 96 });
    expect(toJSON()).toMatchSnapshot();
  });

  it('should render with heading2 variant', () => {
    const { getByTestId, toJSON } = renderWithTheme(<Typography testID='typography' variant='heading2'>Typography</Typography>);

    expect(getByTestId('typography')).toHaveStyle({ fontSize: 60 });
    expect(toJSON()).toMatchSnapshot();
  });

  it('should render with heading3 variant', () => {
    const { getByTestId, toJSON } = renderWithTheme(<Typography testID='typography' variant='heading3'>Typography</Typography>);

    expect(getByTestId('typography')).toHaveStyle({ fontSize: 48 });
    expect(toJSON()).toMatchSnapshot();
  });

  it('should render with heading4 variant', () => {
    const { getByTestId, toJSON } = renderWithTheme(<Typography testID='typography' variant='heading4'>Typography</Typography>);

    expect(getByTestId('typography')).toHaveStyle({ fontSize: 34 });
    expect(toJSON()).toMatchSnapshot();
  });

  it('should render with heading5 variant', () => {
    const { getByTestId, toJSON } = renderWithTheme(<Typography testID='typography' variant='heading5'>Typography</Typography>);

    expect(getByTestId('typography')).toHaveStyle({ fontSize: 24 });
    expect(toJSON()).toMatchSnapshot();
  });

  it('should render with heading6 variant', () => {
    const { getByTestId, toJSON } = renderWithTheme(<Typography testID='typography' variant='heading6'>Typography</Typography>);

    expect(getByTestId('typography')).toHaveStyle({ fontSize: 20 });
    expect(toJSON()).toMatchSnapshot();
  });


  it('should render with body1 variant', () => {
    const { getByTestId, toJSON } = renderWithTheme(<Typography testID='typography' variant='body1'>Typography</Typography>);

    expect(getByTestId('typography')).toHaveStyle({ fontSize: 16 });
    expect(toJSON()).toMatchSnapshot();
  });


  it('should render with body2 variant', () => {
    const { getByTestId, toJSON } = renderWithTheme(<Typography testID='typography' variant='body2'>Typography</Typography>);

    expect(getByTestId('typography')).toHaveStyle({ fontSize: 14 });
    expect(toJSON()).toMatchSnapshot();
  });


  it('should render with caption variant', () => {
    const { getByTestId, toJSON } = renderWithTheme(<Typography testID='typography' variant='caption'>Typography</Typography>);

    expect(getByTestId('typography')).toHaveStyle({ fontSize: 12 });
    expect(toJSON()).toMatchSnapshot();
  });


  it('should render with overline variant', () => {
    const { getByTestId, toJSON } = renderWithTheme(<Typography testID='typography' variant='overline'>Typography</Typography>);

    expect(getByTestId('typography')).toHaveStyle({ fontSize: 12 });
    expect(toJSON()).toMatchSnapshot();
  });
});
