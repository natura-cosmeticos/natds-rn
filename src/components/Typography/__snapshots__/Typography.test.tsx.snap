// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typography Component should render with body1 variant 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#000000",
        "fontSize": 16,
        "fontWeight": "400",
        "letterSpacing": 0.51,
      },
    ]
  }
  testID="typography"
  variant="body1"
>
  Typography
</Text>
`;

exports[`Typography Component should render with body2 variant 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#000000",
        "fontSize": 14,
        "fontWeight": "400",
        "letterSpacing": 0.22,
      },
    ]
  }
  testID="typography"
  variant="body2"
>
  Typography
</Text>
`;

exports[`Typography Component should render with caption variant 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#000000",
        "fontSize": 12,
        "fontWeight": "400",
        "letterSpacing": 0.38,
      },
    ]
  }
  testID="typography"
  variant="caption"
>
  Typography
</Text>
`;

exports[`Typography Component should render with default variant 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#000000",
        "fontSize": 96,
        "fontWeight": "400",
        "letterSpacing": 0,
      },
    ]
  }
  testID="typography"
>
  Typography
</Text>
`;

exports[`Typography Component should render with heading2 variant 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#000000",
        "fontSize": 60,
        "fontWeight": "400",
        "letterSpacing": 0,
      },
    ]
  }
  testID="typography"
  variant="heading2"
>
  Typography
</Text>
`;

exports[`Typography Component should render with heading3 variant 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#000000",
        "fontSize": 48,
        "fontWeight": "400",
        "letterSpacing": 0,
      },
    ]
  }
  testID="typography"
  variant="heading3"
>
  Typography
</Text>
`;

exports[`Typography Component should render with heading4 variant 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#000000",
        "fontSize": 34,
        "fontWeight": "400",
        "letterSpacing": 0.27,
      },
    ]
  }
  testID="typography"
  variant="heading4"
>
  Typography
</Text>
`;

exports[`Typography Component should render with heading5 variant 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#000000",
        "fontSize": 24,
        "fontWeight": "400",
        "letterSpacing": 0,
      },
    ]
  }
  testID="typography"
  variant="heading5"
>
  Typography
</Text>
`;

exports[`Typography Component should render with heading6 variant 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#000000",
        "fontSize": 20,
        "fontWeight": "500",
        "letterSpacing": 0.24,
      },
    ]
  }
  testID="typography"
  variant="heading6"
>
  Typography
</Text>
`;

exports[`Typography Component should render with overline variant 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#000000",
        "fontSize": 12,
        "fontWeight": "500",
        "letterSpacing": 1.92,
      },
    ]
  }
  testID="typography"
  variant="overline"
>
  Typography
</Text>
`;
